Source/Core/VideoCommon/PixelShaderGen.cpp

```diff
@@ -555,7 +565,9 @@ static inline void GeneratePixelShader(T& out, DSTALPHA_MODE dstAlphaMode, API_T
 	}
 
 	if (per_pixel_depth && bpmem.UseLateDepthTest())
-		out.Write("\tdepth = float(zCoord) / float(0xFFFFFF);\n");
+	{
+		WritePerPixelDepth<T>(out, uid_data, ApiType);
```

Comment:
Is there a reason to apply z-freeze after z-textures here?
I think the froozen plane is within the primitive assembly step, so before the hardware rasterization. The z-texture is handled after the TEV, so it likely still apply through z-freeze is in use.

My suggestion is to revert the "WritePerPixelDepth" change and to apply the z-freeze code in: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/PixelShaderGen.cpp#L565
if (zfreeze) {} else if (fast_depth) {} else {}